File: sdk-ts/src/overlay.ts
/**
 * aria-bsv/overlay — BSV Overlay Services client (BRC-31).
 *
 * Mirrors ``aria.overlay`` from the Python SDK.
 *
 * Provides:
 * - {@link TopicManager}: submit transactions to overlay topics.
 * - {@link LookupService}: query outputs by system_id/epoch_id.
 * - {@link OverlayClient}: high-level combined interface.
 *
 * ARIA topics:
 * - ``tm_aria_epochs`` — EPOCH_OPEN and EPOCH_CLOSE transactions.
 * - ``ls_aria``        — Lookup service for audit records.
 */

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface AdmittanceResult {
  /** 64-char hex transaction ID (may be empty if broadcast failed). */
  txid: string;
  /** Topic name. */
  topic: string;
  /** True if at least one output was admitted. */
  admitted: boolean;
  /** 0-based indices of admitted outputs. */
  admittedOutputs: number[];
  /** Human-readable status message. */
  message: string;
}

export interface LookupResult {
  txid: string;
  outputIndex: number;
  beef: string | null;
  data: Record<string, unknown>;
}

export interface OverlayClientOptions {
  baseUrl: string;
  apiKey?: string;
  epochTopic?: string;
  lookupService?: string;
}

// ---------------------------------------------------------------------------
// Fetch abstraction (replaceable in tests)
// ---------------------------------------------------------------------------

type FetchLike = (
  url: string,
  init: { method: string; headers: Record<string, string>; body: string }
) => Promise<{ ok: boolean; status: number; json(): Promise<unknown>; text(): Promise<string> }>;

let _fetchImpl: FetchLike = async (url, init) => {
  if (typeof globalThis.fetch === "function") {
    return globalThis.fetch(url, init);
  }
  if (typeof require !== "undefined") {
    try {
      // eslint-disable-next-line @typescript-eslint/no-var-requires
      const mod = require("node-fetch") as { default: FetchLike };
      return mod.default(url, init);
    } catch {
      // not installed
    }
  }
  throw new Error("No fetch available. Use Node 18+ or install node-fetch.");
};

/** Replace the fetch implementation (useful for testing). */
export function _setOverlayFetchImpl(fn: FetchLike): void {
  _fetchImpl = fn;
}

// ---------------------------------------------------------------------------
// TopicManager
// ---------------------------------------------------------------------------

/**
 * Submit BSV transactions to an overlay topic.
 *
 * @example
 * ```ts
 * const tm = new TopicManager({ baseUrl: "https://overlay.example.com", apiKey: "..." });
 * const result = await tm.submit(rawTxHex);
 * ```
 */
export class TopicManager {
  private readonly _base: string;
  private readonly _topic: string;
  private readonly _apiKey: string | undefined;

  constructor(options: {
    baseUrl: string;
    topic?: string;
    apiKey?: string;
  }) {
    this._base = options.baseUrl.replace(/\/$/, "");
    this._topic = options.topic ?? "tm_aria_epochs";
    this._apiKey = options.apiKey;
  }

  get topic(): string {
    return this._topic;
  }

  /**
   * Submit a raw hex transaction to the topic.
   *
   * @param rawTx Hex-encoded raw transaction.
   * @returns {@link AdmittanceResult} with the overlay node's decision.
   */
  async submit(rawTx: string): Promise<AdmittanceResult> {
    const url = `${this._base}/v1/submit`;
    const body = JSON.stringify({ rawTx, topics: [this._topic] });
    const headers = this._headers();

    const resp = await _fetchImpl(url, { method: "POST", headers, body });
    if (!resp.ok) {
      const text = await resp.text();
      throw new Error(`Overlay node HTTP ${resp.status}: ${text.slice(0, 200)}`);
    }

    const json = await resp.json() as {
      txid?: string;
      topics?: Record<string, { outputsToAdmit?: number[] }>;
    };

    const topicData = json.topics?.[this._topic] ?? {};
    const admittedOutputs: number[] = topicData.outputsToAdmit ?? [];
    const admitted = admittedOutputs.length > 0;

    return {
      txid: json.txid ?? "",
      topic: this._topic,
      admitted,
      admittedOutputs,
      message: admitted ? "OK" : "No outputs admitted",
    };
  }

  private _headers(): Record<string, string> {
    const h: Record<string, string> = {
      "Content-Type": "application/json",
      "Accept": "application/json",
    };
    if (this._apiKey) h["Authorization"] = `Bearer ${this._apiKey}`;
    return h;
  }
}

// ---------------------------------------------------------------------------
// LookupService
// ---------------------------------------------------------------------------

/**
 * Query an overlay lookup service for BSV outputs.
 *
 * @example
 * ```ts
 * const ls = new LookupService({ baseUrl: "https://overlay.example.com" });
 * const results = await ls.lookup({ system_id: "my-system" });
 * ```
 */
export class LookupService {
  private readonly _base: string;
  private readonly _service: string;
  private readonly _apiKey: string | undefined;

  constructor(options: {
    baseUrl: string;
    serviceName?: string;
    apiKey?: string;
  }) {
    this._base = options.baseUrl.replace(/\/$/, "");
    this._service = options.serviceName ?? "ls_aria";
    this._apiKey = options.apiKey;
  }

  get serviceName(): string {
    return this._service;
  }

  /**
   * Query the lookup service.
   *
   * @param query  Service-specific query object.
   * @param limit  Max results (default 10).
   */
  async lookup(
    query: Record<string, unknown>,
    limit = 10
  ): Promise<LookupResult[]> {
    const url = `${this._base}/v1/lookup`;
    const body = JSON.stringify({ service: this._service, query, limit });
    const headers = this._headers();

    const resp = await _fetchImpl(url, { method: "POST", headers, body });
    if (!resp.ok) {
      const text = await resp.text();
      throw new Error(`Lookup service HTTP ${resp.status}: ${text.slice(0, 200)}`);
    }

    const json = await resp.json() as { results?: unknown[] };
    return (json.results ?? []).map((item) => {
      const i = item as Record<string, unknown>;
      return {
        txid: String(i["txid"] ?? ""),
        outputIndex: Number(i["outputIndex"] ?? 0),
        beef: (i["beef"] as string | null | undefined) ?? null,
        data: (i["data"] as Record<string, unknown>) ?? {},
      };
    });
  }

  private _headers(): Record<string, string> {
    const h: Record<string, string> = {
      "Content-Type": "application/json",
      "Accept": "application/json",
    };
    if (this._apiKey) h["Authorization"] = `Bearer ${this._apiKey}`;
    return h;
  }
}

// ---------------------------------------------------------------------------
// OverlayClient
// ---------------------------------------------------------------------------

/**
 * High-level client combining {@link TopicManager} and {@link LookupService}.
 *
 * @example
 * ```ts
 * const client = new OverlayClient({ baseUrl: "https://overlay.example.com", apiKey: "..." });
 * await client.submitEpoch(rawTxHex);
 * const epochs = await client.findEpochs("my-system");
 * ```
 */
export class OverlayClient {
  readonly topicManager: TopicManager;
  readonly lookupService: LookupService;

  constructor(options: OverlayClientOptions) {
    this.topicManager = new TopicManager({
      baseUrl: options.baseUrl,
      topic: options.epochTopic ?? "tm_aria_epochs",
      apiKey: options.apiKey,
    });
    this.lookupService = new LookupService({
      baseUrl: options.baseUrl,
      serviceName: options.lookupService ?? "ls_aria",
      apiKey: options.apiKey,
    });
  }

  /** Submit an epoch transaction (OPEN or CLOSE) to the overlay network. */
  async submitEpoch(rawTx: string): Promise<AdmittanceResult> {
    return this.topicManager.submit(rawTx);
  }

  /**
   * Find epoch records for a system.
   *
   * @param systemId  ARIA system ID.
   * @param epochId   Optional specific epoch ID.
   * @param limit     Max results.
   */
  async findEpochs(
    systemId: string,
    epochId?: string,
    limit = 10
  ): Promise<LookupResult[]> {
    const query: Record<string, unknown> = { system_id: systemId };
    if (epochId !== undefined) query["epoch_id"] = epochId;
    return this.lookupService.lookup(query, limit);
  }

  /**
   * Find audit records for a specific epoch.
   */
  async findRecords(
    systemId: string,
    epochId: string,
    limit = 100
  ): Promise<LookupResult[]> {
    return this.lookupService.lookup(
      { system_id: systemId, epoch_id: epochId, type: "AUDIT_RECORD" },
      limit
    );
  }
}

File: aria/overlay.py
"""aria.overlay — BSV Overlay Services client (BRC-31).

Provides:
    - TopicManager: submit transactions to a topic and track admitted outputs.
    - LookupService: query a lookup service for outputs matching a given query.
    - OverlayClient: high-level client that combines both.

All network I/O is async (httpx).  Classes are fully dependency-injectable
for testing via the ``client`` parameter.

BRC-31 reference: https://github.com/bitcoin-sv/BRCs/blob/master/overlay/0031.md

The ARIA overlay protocol uses two topics:
    - ``tm_aria_epochs``:  EPOCH_OPEN and EPOCH_CLOSE transactions.
    - ``tm_aria_records``: per-record OP_RETURN anchoring (when enabled).
"""

from __future__ import annotations

import asyncio
import json
from dataclasses import dataclass, field
from typing import Any, Sequence

import httpx

from aria.core.errors import ARIAError

# ---------------------------------------------------------------------------
# Errors
# ---------------------------------------------------------------------------


class OverlayError(ARIAError):
    """Raised when an overlay network operation fails."""


# ---------------------------------------------------------------------------
# Data classes
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class AdmittanceResult:
    """Result of submitting a transaction to a topic.

    Attributes:
        txid:            64-char hex transaction ID.
        topic:           Topic name (e.g. ``"tm_aria_epochs"``).
        admitted:        True if at least one output was admitted.
        admitted_outputs: 0-based indices of admitted outputs.
        message:         Human-readable status message.
    """

    txid: str
    topic: str
    admitted: bool
    admitted_outputs: list[int]
    message: str


@dataclass(frozen=True)
class LookupResult:
    """A single output returned by a lookup service query.

    Attributes:
        txid:       Transaction ID.
        output_index: Output index within the transaction.
        beef:       BEEF-encoded transaction data (hex), if available.
        data:       Arbitrary metadata returned by the lookup service.
    """

    txid: str
    output_index: int
    beef: str | None
    data: dict[str, Any]


# ---------------------------------------------------------------------------
# TopicManager
# ---------------------------------------------------------------------------


class TopicManager:
    """Submit BSV transactions to an overlay topic.

    The topic manager admits (or rejects) transactions based on protocol
    rules.  For ARIA, epoch transactions are submitted to ``tm_aria_epochs``.

    Args:
        base_url: Overlay node base URL.
        topic:    Topic name (default: ``"tm_aria_epochs"``).
        api_key:  Optional Bearer token for the overlay node.
        client:   Optional ``httpx.AsyncClient`` (injected in tests).
    """

    def __init__(
        self,
        base_url: str,
        topic: str = "tm_aria_epochs",
        api_key: str | None = None,
        client: httpx.AsyncClient | None = None,
    ) -> None:
        self._base = base_url.rstrip("/")
        self._topic = topic
        self._api_key = api_key
        self._client = client

    @property
    def topic(self) -> str:
        return self._topic

    async def submit(self, raw_tx: str) -> AdmittanceResult:
        """Submit a raw hex transaction to the topic.

        Args:
            raw_tx: Hex-encoded raw transaction bytes.

        Returns:
            :class:`AdmittanceResult` describing the overlay node's decision.

        Raises:
            OverlayError: On network failure or unexpected HTTP error.
        """
        url = f"{self._base}/v1/submit"
        headers = self._headers()
        body = {"rawTx": raw_tx, "topics": [self._topic]}

        try:
            resp = await self._get(url, headers=headers, body=body)
        except OverlayError:
            raise
        except Exception as exc:
            raise OverlayError(f"Failed to submit to topic {self._topic}: {exc}") from exc

        # BRC-31 response: {"topics": {"tm_aria_epochs": {"outputsToAdmit": [...], "coinsToRetain": [...]}}}
        topic_data = resp.get("topics", {}).get(self._topic, {})
        admitted_outputs: list[int] = topic_data.get("outputsToAdmit", [])
        admitted = len(admitted_outputs) > 0

        # Extract txid from BEEF or response
        txid: str = resp.get("txid", "")

        return AdmittanceResult(
            txid=txid,
            topic=self._topic,
            admitted=admitted,
            admitted_outputs=admitted_outputs,
            message="OK" if admitted else "No outputs admitted",
        )

    async def _get(
        self,
        url: str,
        headers: dict[str, str],
        body: dict[str, Any],
    ) -> dict[str, Any]:
        if self._client is not None:
            resp = await self._client.post(url, json=body, headers=headers, timeout=15.0)
        else:
            async with httpx.AsyncClient() as c:
                resp = await c.post(url, json=body, headers=headers, timeout=15.0)
        if not resp.is_success:
            raise OverlayError(
                f"Overlay node returned HTTP {resp.status_code}: {resp.text[:200]}"
            )
        return resp.json()  # type: ignore[no-any-return]

    def _headers(self) -> dict[str, str]:
        h: dict[str, str] = {
            "Content-Type": "application/json",
            "Accept": "application/json",
        }
        if self._api_key:
            h["Authorization"] = f"Bearer {self._api_key}"
        return h


# ---------------------------------------------------------------------------
# LookupService
# ---------------------------------------------------------------------------


class LookupService:
    """Query an overlay lookup service for outputs.

    Args:
        base_url:        Overlay node base URL.
        service_name:    Lookup service name (default: ``"ls_aria"``).
        api_key:         Optional Bearer token.
        client:          Optional ``httpx.AsyncClient`` (injected in tests).
    """

    def __init__(
        self,
        base_url: str,
        service_name: str = "ls_aria",
        api_key: str | None = None,
        client: httpx.AsyncClient | None = None,
    ) -> None:
        self._base = base_url.rstrip("/")
        self._service = service_name
        self._api_key = api_key
        self._client = client

    @property
    def service_name(self) -> str:
        return self._service

    async def lookup(
        self,
        query: dict[str, Any],
        limit: int = 10,
    ) -> list[LookupResult]:
        """Query the lookup service for matching outputs.

        Args:
            query:  Service-specific query dict.  For ARIA, pass e.g.
                    ``{"system_id": "my-system", "epoch_id": "..."}``
            limit:  Maximum number of results to return (default: 10).

        Returns:
            List of :class:`LookupResult` objects.

        Raises:
            OverlayError: On network failure.
        """
        url = f"{self._base}/v1/lookup"
        headers = self._headers()
        body = {
            "service": self._service,
            "query": query,
            "limit": limit,
        }

        try:
            raw = await self._post(url, headers=headers, body=body)
        except OverlayError:
            raise
        except Exception as exc:
            raise OverlayError(
                f"Lookup service {self._service} query failed: {exc}"
            ) from exc

        results: list[LookupResult] = []
        for item in raw.get("results", []):
            results.append(
                LookupResult(
                    txid=str(item.get("txid", "")),
                    output_index=int(item.get("outputIndex", 0)),
                    beef=item.get("beef"),
                    data=dict(item.get("data", {})),
                )
            )
        return results

    async def _post(
        self,
        url: str,
        headers: dict[str, str],
        body: dict[str, Any],
    ) -> dict[str, Any]:
        if self._client is not None:
            resp = await self._client.post(url, json=body, headers=headers, timeout=15.0)
        else:
            async with httpx.AsyncClient() as c:
                resp = await c.post(url, json=body, headers=headers, timeout=15.0)
        if not resp.is_success:
            raise OverlayError(
                f"Lookup service returned HTTP {resp.status_code}: {resp.text[:200]}"
            )
        return resp.json()  # type: ignore[no-any-return]

    def _headers(self) -> dict[str, str]:
        h: dict[str, str] = {
            "Content-Type": "application/json",
            "Accept": "application/json",
        }
        if self._api_key:
            h["Authorization"] = f"Bearer {self._api_key}"
        return h


# ---------------------------------------------------------------------------
# OverlayClient — high-level combined interface
# ---------------------------------------------------------------------------


class OverlayClient:
    """High-level client combining a :class:`TopicManager` and a
    :class:`LookupService` for the ARIA overlay protocol.

    Args:
        base_url:       Overlay node base URL.
        epoch_topic:    Topic for epoch transactions
                        (default: ``"tm_aria_epochs"``).
        lookup_service: Lookup service name (default: ``"ls_aria"``).
        api_key:        Optional Bearer token.
        client:         Optional ``httpx.AsyncClient``.
    """

    def __init__(
        self,
        base_url: str,
        epoch_topic: str = "tm_aria_epochs",
        lookup_service: str = "ls_aria",
        api_key: str | None = None,
        client: httpx.AsyncClient | None = None,
    ) -> None:
        self.topic_manager = TopicManager(
            base_url=base_url,
            topic=epoch_topic,
            api_key=api_key,
            client=client,
        )
        self.lookup_service = LookupService(
            base_url=base_url,
            service_name=lookup_service,
            api_key=api_key,
            client=client,
        )

    async def submit_epoch(self, raw_tx: str) -> AdmittanceResult:
        """Submit an epoch transaction (OPEN or CLOSE) to the overlay network."""
        return await self.topic_manager.submit(raw_tx)

    async def find_epochs(
        self,
        system_id: str,
        epoch_id: str | None = None,
        limit: int = 10,
    ) -> list[LookupResult]:
        """Query the overlay network for epoch records.

        Args:
            system_id:  ARIA system ID to filter by.
            epoch_id:   Optional specific epoch ID.
            limit:      Maximum results to return.
        """
        query: dict[str, Any] = {"system_id": system_id}
        if epoch_id is not None:
            query["epoch_id"] = epoch_id
        return await self.lookup_service.lookup(query, limit=limit)

    async def find_records(
        self,
        system_id: str,
        epoch_id: str,
        limit: int = 100,
    ) -> list[LookupResult]:
        """Query the overlay network for audit records in an epoch."""
        query = {
            "system_id": system_id,
            "epoch_id": epoch_id,
            "type": "AUDIT_RECORD",
        }
        return await self.lookup_service.lookup(query, limit=limit)
